### PR TITLE
376-Search in the applicant listing done

### PR DESCRIPTION
**Pull request by Satendra-SR**
#376 
1. In ApplicationController
* checked for the request.
   * if there is a request then `if `section will be executed otherwise `else` will be executed.
   * filters is applied for different job type
2. In application/index.blade.php
  * A search form is added
  * Current `status `is given with as `<input type ="hidden">`
  * Search will be performed for the same status only
  * (Clear current search and filters)` is provided for clear the searched data

### CODE REVIEW

**Comment by bohraji** [comment]
@satendra95 Please request review from @nishanth-kd as discussed yesterday. Also, in the for the pull request above ensure the following information is present-

- The original issue is mentioned, so the reviewer can see the original issue.
- Mention what does this PR does - what functionality it adds, what it removes, etc.
- No need to mention code snippets in description. The reviewer will see it in the code section.

Please update the description before requesting review.

File: app/Http/Controllers/HR/Applications/ApplicationController.php [hunk 1]
```diff
@@ -29,29 +29,49 @@ public function index()
         $filters = [
             'status' => request()->get('status') ?: 'non-rejected',
             'job-type' => $this->getApplicationType(),
-            'job' => request()->get('hr_job_id')
+            'job' => request()->get('hr_job_id'),
         ];
+        if (request()->has('search')) {
+            $search = request()->get('search');
+            $applications = Application::with(['applicant', 'job'])
+                ->whereHas('applicant', function ($query) use ($search) {
```
Comment (nishanth-kd) [inline app/Http/Controllers/HR/Applications/ApplicationController.php]: this should go into the model `Application`.

File: resources/views/hr/application/index.blade.php [hunk 2]
```diff
@@ -5,7 +5,35 @@
...
     @include('hr.menu')
     <br><br>
-    <h1>Applications</h1>
+    <div class="row">
+        <div class="col-md-3">
+            <h1>Applications</h1>
+        </div>
+        <div class="col-md-9">
+            <form class="form-inline" method="GET" action="/{{ Request::path() }}">
+            @if(request('status')=='on-hold')
```
Comment (nishanth-kd) [inline resources/views/hr/application/index.blade.php]: Learn how to build a blade directive for switch case and use it here. 

https://murze.be/building-a-switch-blade-directive by @freekmurze

File: resources/views/hr/application/index.blade.php [hunk 3]
```diff
@@ -5,7 +5,35 @@
...
+                <input type="hidden" name="status" class="form-control" id="search" value="{{ config('constants.hr.status.on-hold.label') }}">
+            @elseif(request('status')=='rejected')
+                <input type="hidden" name="status" class="form-control" id="search" value="{{ config('constants.hr.status.rejected.label') }}">
+            @elseif(request('status')=='no-show')
+                <input type="hidden" name="status" class="form-control" id="search" value="{{ config('constants.hr.status.no-show.label') }}">
```
Comment (nishanth-kd) [inline resources/views/hr/application/index.blade.php]: `<input type="hidden" name="status" class="form-control" id="search" value="` can go outside.

File: resources/views/hr/application/index.blade.php [hunk 4]
```diff
@@ -5,7 +5,35 @@
...
+            @endif
+               <div class="form-group">
+                   <input type="text" name="search" class="form-control" id="search" placeholder="search applicants">
+               </div>
+               <button class="btn btn-info ml-2">Search</button>
+            </form>
+        </div>
+    </div>
+    @if(request()->has('search'))
```
Comment (nishanth-kd) [inline resources/views/hr/application/index.blade.php]: what about filters ? Once you move search to filters[], check if filters all are empty or any single one is present.

File: resources/views/hr/application/index.blade.php [hunk 5]
```diff
@@ -5,7 +5,35 @@
...
+            <form class="form-inline" method="GET" action="/{{ Request::path() }}">
+            @if(request('status')=='on-hold')
+                <input type="hidden" name="status" class="form-control" id="search" value="{{ config('constants.hr.status.on-hold.label') }}">
+            @elseif(request('status')=='rejected')
+                <input type="hidden" name="status" class="form-control" id="search" value="{{ config('constants.hr.status.rejected.label') }}">
+            @elseif(request('status')=='no-show')
+                <input type="hidden" name="status" class="form-control" id="search" value="{{ config('constants.hr.status.no-show.label') }}">
+            @endif
+               <div class="form-group">
+                   <input type="text" name="search" class="form-control" id="search" placeholder="search applicants">
```
Comment (nishanth-kd) [inline resources/views/hr/application/index.blade.php]: casing of `search applicants` ? Is this supposed to be sentence case or lower case?

**Comment by Satendra-SR** [comment]
@nishanth-kd 
done with feedback, request for review
>what I have changed is:
* uses **switch blade directive** instead of **if-else** in `application/index.blade.php`.
* code for search functionality is written inside the **Application model**.
* Minimized code duplication in `ApplicationController` and `view`  also.
* `Search applicants` is written is sentence cases.
* filters is working after `clearing` search record.

**Comment by Satendra-SR** [comment]
> what I improved
* code readability is improved
* code duplicacy is resolved in `application's index` page

**Comment by Satendra-SR** [comment]
>**Changes:**
* _UI_ was broken, made changes to improve _UI_ in `application/index.blade.php`

File: app/Http/Controllers/HR/Applications/ApplicationController.php [hunk 6]
```diff
@@ -31,37 +31,19 @@ public function index()
...
-                [
-                'applications' => $applications,
-                'status' => request()->get('status'),
-            ];
-
-        }
+        $search = request()->get('search');
+        $applications = Application::with(['applicant', 'job'])
+            ->applyFilter($filters)
+            ->byName($search)
```
Comment (nishanth-kd) [inline app/Http/Controllers/HR/Applications/ApplicationController.php]: Move this to `applyFilter`

File: app/Models/HR/Application.php [hunk 7]
```diff
@@ -134,6 +134,15 @@ public function scopeFilterByJobType($query, $type)
         return $query;
     }
 
+    public function scopeByName($query, $search)
```
Comment (nishanth-kd) [inline app/Models/HR/Application.php]: it should either by 'scopeFilterByName' or we should change the others to not have `Filter` for example `FilterByJobType` => 'byJobType'.

I think let's keep it as FilterByName.

**Comment by Satendra-SR** [comment]
> **done with feedback**
* moved `request()->get('search)` inside the filters
* search is happening using `FilterByName` filter

**Comment by nishanth-kd** [comment]
Alright looks good now. 

- [ ] For duplication.
- [ ] For performance. 
- [ ] Usage of recursive methods, loops and algorithms.
- [ ] For readability and understanding. 
- [ ] For functions with multiple responsibilities.
- [ ] For scope of refactoring. 
- [ ] For coupling
- [ ] For alternative approaches to the implementation. 
- [ ] For how the change impacts other parts of the system.
- [ ] For validations. client side and server side validations are present.

For more insights on code reviews checkout our [guidelines](https://docs.google.com/document/d/1Sa-XYJEmGwrOM5gk5zoan_Z6j90ylibmCmt331361nw/edit#).

**Pull request merged**